{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filtering MAF files\n",
    "The current filtering of MAF files has four parts: the quorum rule, filter, gdc filter, and tumor_normal pairs. The quorum rule can easily be hanged by replacing the hard coded 2 with any different desired quorum rule. Because the filtering is done with a pandas dataframe, the filtering is generally straightforward, aside from the quorum and tumor-normal pairs filtering. For example, if you want to filter out any variants that do or do not have certain values in specific columns of the MAF file, only two lines of code is needed:  \n",
    "\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "final = final.drop(index = [i for i in final.index if final.iloc[i][110] != 'PASS'])  \n",
    "final.index = range(len(final))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is also important to note that two additional columns have been added to this filtered MAF file: a list of the callers that had called that variant, and the caller from which the line denoting that variant came from. Because of this, if it is important to consider the values of all of these columns that are different across the MAF files, this information can be recovered with these two extra columns as well as the original MAF files.   \n",
    "Both cells below must be run in order to generate and save these filtered MAF files. The name of the resulting sav"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import math\n",
    "\n",
    "\n",
    "# parse comments from MAF files\n",
    "def parse_comments(fp):\n",
    "    comments = []\n",
    "\n",
    "    fh = open(fp, 'r')\n",
    "    for line in fh:\n",
    "        if line[0] == '#':\n",
    "            comments.append(line)\n",
    "\n",
    "    return(comments)\n",
    "\n",
    "# parses chromosome, start position, end position, reference allele, and tumor alleles of variants\n",
    "def parse_maf(fp):\n",
    "    ''' Return a dictionary with vars[patient][variant_set]'''\n",
    "    caller_vars_dict = {}\n",
    "\n",
    "    index = 0\n",
    "    fh = open(fp, 'r')\n",
    "    for line in fh:\n",
    "        temp = line.strip().split('\\t')\n",
    "        if line[0] == '#' or temp[0] == 'Hugo_Symbol':\n",
    "            continue\n",
    "\n",
    "        chrom =temp[4]\n",
    "        s_pos = temp[5]\n",
    "        e_pos = temp[6]\n",
    "        ref = temp[10]\n",
    "        tum_allele_1 = temp[11]\n",
    "        tum_allele_2 = temp[12]\n",
    "\n",
    "        var = (chrom, s_pos, e_pos, ref, tum_allele_1, tum_allele_2)\n",
    "        \n",
    "        caller_vars_dict[var] = index\n",
    "\n",
    "        index += 1\n",
    "\n",
    "    return caller_vars_dict\n",
    "\n",
    "\n",
    "possible_callers = ('muse', 'mutect', 'somaticsniper', 'varscan')\n",
    "\n",
    "possible_cancers = ('ACC', 'BLCA','BRCA','CESC','CHOL','COAD','DLBC','ESCA','GBM','HNSC', \n",
    "                    'KICH','KIRC','KIRP','LAML','LGG','LIHC','LUAD','LUSC','MESO', 'OV','PAAD',\n",
    "                    'PCPG','PRAD','READ','SARC','SKCM','STAD','TGCT','THCA','THYM','UCEC','UCS','UVM')\n",
    "\n",
    "cwd = os.getcwd()\n",
    "\n",
    "dataframes = {}\n",
    "\n",
    "for cancer in possible_cancers:\n",
    "    # accessing MAF files in path cwd/data/cancer\n",
    "    mafs = os.listdir(os.path.join(cwd, 'data', cancer))\n",
    "\n",
    "    maf_fps = {}\n",
    "    for caller in possible_callers:\n",
    "        for maf in mafs:    \n",
    "            if caller in maf: # str in the filepath\n",
    "                maf_fps[caller] = os.path.join(cwd, 'data', cancer, maf)\n",
    "    \n",
    "    # deletes comments from MAF files and stores them in all_comments         \n",
    "    all_comments = {}\n",
    "    for caller in maf_fps:\n",
    "        all_comments[caller] = parse_comments(maf_fps[caller])\n",
    "        for i in range(4):\n",
    "            del all_comments[caller][0]\n",
    "        all_comments[caller] = all_comments[caller][0].strip().split(',')\n",
    "        all_comments[caller][0] = all_comments[caller][0].strip().split(' ')\n",
    "        all_comments[caller].append(all_comments[caller][0][1])\n",
    "        del all_comments[caller][0]\n",
    "    \n",
    "    # stores MAF files as dataframes and subvariants from parse_maf function for all variants from all callers\n",
    "    all_variants = {}\n",
    "    all_df = {}\n",
    "    for caller in maf_fps:\n",
    "        all_variants[caller] = parse_maf(maf_fps[caller])\n",
    "        all_df[caller] = pd.read_csv(maf_fps[caller], sep='\\t',comment=\"#\")\n",
    "    \n",
    "    # list of all variants\n",
    "    all_v = []\n",
    "    \n",
    "    for caller in maf_fps:\n",
    "        all_v.extend(all_variants[caller].keys())   \n",
    "    \n",
    "    # removes duplicates\n",
    "    all_v = list(dict.fromkeys(all_v))\n",
    "    \n",
    "    # stores indexes and caller of each sub-variant\n",
    "    all_v_index = {k:[] for k in all_v}\n",
    "    all_v_caller = {j:[] for j in all_v}\n",
    "    for caller in maf_fps:\n",
    "        for var in all_v:\n",
    "            if var in all_variants[caller].keys():\n",
    "                all_v_index[var].append(all_variants[caller][var])\n",
    "                all_v_caller[var].append(caller)\n",
    "    \n",
    "    # quorum rule\n",
    "    list_delete=[]\n",
    "    for i in all_v_index.keys():\n",
    "        if len(all_v_index[i]) < 2:\n",
    "            list_delete.append(i)\n",
    "    \n",
    "    for j in list_delete:        \n",
    "        del all_v_index[j]\n",
    "        del all_v_caller[j]\n",
    "    \n",
    "    # gathers which variants from which caller to include in final file\n",
    "    \n",
    "    boolean_arrays = {c:[] for c in possible_callers}\n",
    "    for c in possible_callers:\n",
    "        boolean_arrays[c] = np.zeros(len(all_df[c]), dtype=bool)\n",
    "    \n",
    "    orig = []\n",
    "    for k in all_v_index.keys():\n",
    "        boolean_arrays[all_v_caller[k][0]][all_v_index[k][0]] = True\n",
    "        orig.append(all_v_caller[k][0])\n",
    "    \n",
    "    final_dataframes = {}\n",
    "    for c in possible_callers:\n",
    "        final_dataframes[c] = all_df[c][boolean_arrays[c]]\n",
    "        \n",
    "    final = pd.concat(final_dataframes, ignore_index = True, keys= None)\n",
    "    \n",
    "    final['callers'] = list(all_v_caller.values())\n",
    "    final['original'] = orig\n",
    "    \n",
    "    # PASS filter and reindexing\n",
    "    final = final.drop(index = [i for i in final.index if final.iloc[i][110] != 'PASS'])\n",
    "    final.index = range(len(final))\n",
    "    \n",
    "    #gdc filter and reindexing \n",
    "    gdc_filt = set(['common_in_exac', 'gdc_pon', 'ndp', 'NonExonic', 'bitgt'])\n",
    "    gdc_filt_drop = []\n",
    "    for i in final.index:\n",
    "        if isinstance(final.iloc[i][116], str):\n",
    "            if set(final.iloc[i][116].split(';')) & gdc_filt != set():\n",
    "                gdc_filt_drop.append(i)\n",
    "    final = final.drop(index = gdc_filt_drop)\n",
    "    final.index = range(len(final))\n",
    "    \n",
    "    \n",
    "    # filter out any variants whose tumor-normal pair was run through all 4 variant callers\n",
    "    tumor_normal = {c:[] for caller in maf_fps}\n",
    "    \n",
    "    for caller in maf_fps:    \n",
    "        tumor_normal[caller]= [all_df[caller].loc[:,'Tumor_Sample_Barcode'],all_df[caller].loc[:,'Matched_Norm_Sample_Barcode']]\n",
    "        tumor_normal[caller] = np.array(tumor_normal[caller])\n",
    "        tumor_normal[caller] = tumor_normal[caller].transpose()\n",
    "        tumor_normal[caller] = set([(tumor_normal[caller][i][0],tumor_normal[caller][i][1]) for i in range(len(all_df[caller]))])\n",
    "    \n",
    "    tumor_normal_set = tumor_normal[list(maf_fps.keys())[0]]\n",
    "\n",
    "    for caller in maf_fps:\n",
    "        tumor_normal_set = tumor_normal_set & tumor_normal[caller]\n",
    "        \n",
    "    final = final.drop(index = [i for i in final.index if (final.iloc[i][15], final.iloc[i][16]) not in tumor_normal_set])\n",
    "    \n",
    "    final.index = range(len(final))\n",
    "    \n",
    "    # store in dictionary of dataframes\n",
    "    dataframes[cancer] = final\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cancer in possible_cancers:\n",
    "    dataframes[cancer].to_csv(\"filtered_{}.maf\".format(cancer), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
